src/extra/gfx/main_window.cpp

```diff
@@ -187,6 +193,60 @@ struct MainWindow final {
 
   void frameBegan() { _graphicsContext.getDrawQueue()->clear(); }
 
+  void updateVirtualInputEvents() {
+    auto &virtualInputEvents = _inputContext.virtualInputEvents;
+    auto &heldKeys = _inputContext.heldKeys;
+    virtualInputEvents.clear();
+
+    for (auto &evt : _inputContext.events) {
+      switch (evt.type) {
+      case SDL_KEYDOWN:
+        if (heldKeys.insert(evt.key.keysym.sym).second) {
+          virtualInputEvents.push_back(evt);
+        }
+        break;
+      case SDL_KEYUP:
+        if (heldKeys.contains(evt.key.keysym.sym)) {
+          heldKeys.erase(evt.key.keysym.sym);
+          virtualInputEvents.push_back(evt);
+        }
+        break;
+      }
+    }
+
+    int numKeys{};
+    auto keyStates = SDL_GetKeyboardState(&numKeys);
+    _actuallyHeldKeys.clear();
+    for (size_t i = 0; i < numKeys; i++) {
+      SDL_Keycode code = SDL_GetKeyFromScancode((SDL_Scancode)i);
+      if (keyStates[i] == 1) {
+        _actuallyHeldKeys.insert(code);
+      }
+    }
+
+    // Synthesize release events in case the event got lost
+    for (auto it = heldKeys.begin(); it != heldKeys.end();) {
+      if (!_actuallyHeldKeys.contains(*it)) {
+        it = heldKeys.erase(it);
+        // Generate virtual event
+        SDL_Event evt{
+            .key =
+                {
+                    .type = SDL_KEYUP,
+                    .state = SDL_RELEASED,
+                    .keysym =
+                        {
+                            .sym = *it,
+                        },
+                },
+        };

```

Comment:
Weird formatting. Shouldn't it have 2-space ident?

Reply:
Yeah it's what the auto-format did though.